### PR TITLE
[FLINK-16057] Optimize ContinuousFileReaderOperator

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

Migrating `ContinuousFileReaderOperator` to mailbox execution model caused performance regression.
This PR eliminates it by avoiding going through the mailbox unnecessarily.

Local results before and after optimization (with [benchmark fix](https://github.com/dataArtisans/flink-benchmarks/pull/63) applied):
```
ContinuousFileReaderOperatorBenchmark.readFileSplit  thrpt   30  10444.184 ± 228.434  ops/ms
ContinuousFileReaderOperatorBenchmark.readFileSplit  thrpt   30  31451.848 ± 849.005  ops/ms
```
Without the fix, results are better because it then does all the job in CLOSING phase in a loop. 
So it's preferable to merge [benchmark fix](https://github.com/dataArtisans/flink-benchmarks/pull/63) first.

## Brief change log

  - loop in `ContinuousFileReaderOperator` if mailbox executor is idle
  - optimize `MailboxExecutor.isIdle`

## Verifying this change

  - correctness is covered by existing tests such as `ContinuousFileProcessingTest`
  - performance if measured by http://codespeed.dak8s.net:8000/

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java [hunk 1]
```diff
@@ -43,10 +43,22 @@
 
 	private final StreamTaskActionExecutor actionExecutor;
 
+	private MailboxProcessor mailboxProcessor;
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java:46]: `@Nullable` - but check other comments first (for a better solution)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java [hunk 2]
```diff
@@ -43,10 +43,22 @@
...
+	public MailboxExecutorImpl(@Nonnull TaskMailbox mailbox, int priority, StreamTaskActionExecutor actionExecutor, MailboxProcessor mailboxProcessor) {
 		this.mailbox = mailbox;
 		this.priority = priority;
 		this.actionExecutor = Preconditions.checkNotNull(actionExecutor);
+		this.mailboxProcessor = mailboxProcessor;
+	}
+
+	public boolean isIdle() {
+		return !mailboxProcessor.isMailboxLoopRunning() ||
+			(mailboxProcessor.isDefaultActionUnavailable() && !mailbox.hasMail() && mailbox.getState().isAcceptingMails());
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java:61]: `mailboxProcessor != null && mailboxProcessor.isDefaultActionUnavailable()`?  - but check other comments first (for a better solution)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java [hunk 3]
```diff
@@ -114,6 +114,7 @@ public MailboxProcessor(
 		this.actionExecutor = Preconditions.checkNotNull(actionExecutor);
 		this.mailbox = Preconditions.checkNotNull(mailbox);
 		this.mainMailboxExecutor = Preconditions.checkNotNull(mainMailboxExecutor);
+		((MailboxExecutorImpl) this.mainMailboxExecutor).setMailboxProcessor(this);
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java:117]: Couple of problems:
1. it changes an object passed from the outside
2. requires cast (could be solved by changing the signature of the constructor)

I'm thinking that maybe a better solution is to get rid of the `mainMailboxExecutor` and replace getter `getMainMailboxExecutor` with:
```
	public MailboxExecutor getMainMailboxExecutor() {
		return new MailboxExecutorImpl(mailbox, MIN_PRIORITY, actionExecutor, this);
	}
```
+
cache access to `mailboxProcessor.getMainMailboxExecutor()` inside `StreamTask` by creating a field in ` StreamTask`: `private final MailboxExecutor mainMailboxExecutor`?

This would solve the problem of circular dependency and nullable non final field inside `MailboxExecutorImpl`.

WDYT?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 4]
```diff
@@ -311,17 +312,19 @@ private void enqueueProcessRecord() {
...
-		if (format.reachedEnd()) {
-			onSplitProcessed();
-		} else {
-			enqueueProcessRecord();
-		}
+			if (format.reachedEnd()) {
+				onSplitProcessed();
+				return;
+			}
+		} while (executor.isIdle()); // todo: consider moving this loop into MailboxProcessor (return boolean "re-execute" from enqueued action)
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java:326]: nit: I would be slightly in favour of adding this method to the `MailboxExecutor` interface, but if you think opposite let it be for now.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java [hunk 5]
```diff
@@ -50,7 +50,7 @@ public MailboxExecutorImpl(@Nonnull TaskMailbox mailbox, int priority, StreamTas
 	}
 
 	public MailboxExecutorImpl(@Nonnull TaskMailbox mailbox, int priority, StreamTaskActionExecutor actionExecutor, MailboxProcessor mailboxProcessor) {
-		this.mailbox = mailbox;
+		this.mailbox = (TaskMailboxImpl) mailbox;
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java:53]: revert this change?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 6]
```diff
@@ -276,6 +281,7 @@ protected StreamTask(
 		this.actionExecutor = Preconditions.checkNotNull(actionExecutor);
 		this.mailboxProcessor = new MailboxProcessor(this::processInput, mailbox, actionExecutor);
 		this.mailboxProcessor.initMetric(environment.getMetricGroup());
+		this.mainMailboxExecutor = new MailboxExecutorImpl(mailbox, MIN_PRIORITY, actionExecutor);
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:284]: I was thinking about doing here:
```
this.mainMailboxExecutor = mailboxProcessor.getMainMailboxExecutor();
```
or
```
this.mainMailboxExecutor = mailboxProcessor.getMinPriorityMailboxExecutor();
```
to maintain the contract, that everything goes via `MailboxProcessor` one way or another.

As it is now, some actions are going via `mailbox` and `actionExecutor` bypassing the processor, which I'm not sure if it's as clean?

Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java]: Done.
